File: leetcodeContest/186周/1423. 可获得的最大点数.js
/**
 * @param {number[]} cardPoints
 * @param {number} k
 * @return {number}
 */
var maxScore = function (cardPoints, k) {
  let len = cardPoints.length,
    sum = 0
  for (let cardPoint of cardPoints) {
    sum += cardPoint
  }
  let min = Number.MAX_SAFE_INTEGER
  let length = len - k
  let temp = 0
  for (let i = 0; i < len; i++) {
    temp += cardPoints[i]
    if (i >= length) {
      temp -= cardPoints[i - length]
    }
    if (i >= length - 1) {
      min = Math.min(min, temp)
    }
  }
  // 总和减去中间窗口的最小值
  return sum - min
}

File: leetcodeContest/186周/1423. 可获得的最大点数.py
class Solution:
    def maxScore(self, cardPoints: List[int], k: int) -> int:
      if len(cardPoints) == k:
        return (sum(cardPoints))
      else:
        l = len(cardPoints) - k
        left = 0
        right = len(cardPoints) - k
        sum1 = sum(cardPoints[left:right])
        mi = sum1
        for i in range(k):
          re = sum1 - cardPoints[i] + cardPoints[i + l]
          sum1 = re
          if re < mi:
            mi = re
        return sum(cardPoints) - mi
